### PR TITLE
lazy load symbol names indices

### PR DESCRIPTION
**Pull request by kzh**
This PR avoids immediately reading in a `[]uint32` (symbol names indices) from mmap'ed memory. Instead, it keeps the slice as a `[]byte` and converts to `uint32`s when needed. It also disables ranking by symbol boundaries since this is an intensive operation that we don't have a use for. It also includes a bunch of minor changes.

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -95,7 +95,7 @@ type LineFragmentMatch struct {
 	// Number bytes that match.
 	MatchLength int
 
-	SymbolInfo *Symbol
+	SymbolInfo Symbol
```
Comment (keegancsmith) [inline api.go]: why this change? For text search this will always be nil (the more common case).

File: contentprovider.go [hunk 2]
```diff
@@ -255,7 +254,7 @@ func findSection(secs []DocumentSection, off, sz uint32) *DocumentSection {
 	return nil
 }
 
-func matchScore(secs []DocumentSection, m *LineMatch) float64 {
+func matchScore(m *LineMatch) float64 {
```
Comment (keegancsmith) [inline contentprovider.go]: don't change this function, lets just keep it the same as upstream to simplify merging. Rather just always pass nil in for secs.

File: indexdata.go [hunk 3]
```diff
@@ -94,32 +94,39 @@ type symbolData struct {
...
 	symKindIndex   []uint32
-	//symMetadata is [4]uint32 Sym Kind Parent ParentKind
+	// symMetadata is [4]uint32 Sym Kind Parent ParentKind
 	symMetaData []byte
 }
 
+func (d *symbolData) parent(i uint32) []byte {
+	delta := binary.BigEndian.Uint32(d.symIndex)
+	start := binary.BigEndian.Uint32(d.symIndex[i*4:]) - delta
+	end := binary.BigEndian.Uint32(d.symIndex[(i+1)*4:]) - delta
```
Comment (keegancsmith) [inline indexdata.go]: I don't really understand this delta part. Can you explain?

Comment (kzh) [reply indexdata.go]: `parent` is the lazy version of `relativeIndex`.
```go
// relativeIndex returns the relative offsets of the items (first
// element is 0), plus a final marking the end of the last item.
func (s *compoundSection) relativeIndex() []uint32 {
	ri := make([]uint32, 0, len(s.offsets)+1)
	for _, o := range s.offsets {
		ri = append(ri, o-s.offsets[0])
	}
	if len(s.offsets) > 0 {
		ri = append(ri, s.data.sz)
	}
	return ri
}
```
Zoekt stores the absolute offsets for each simple section in a compound section, so subtracting the offset of the first element (delta) from an element's offset will give us the relative offset which we use to access the slice.

**Comment by keegancsmith** [comment]
@kzh please ping me on slack when you want me to review this again.

File: contentprovider.go [hunk 4]
```diff
@@ -154,9 +154,9 @@ func (p *contentProvider) fillMatches(ms []*candidateMatch) []LineMatch {
 		result = p.fillContentMatches(ms)
 	}
 
-	sects := p.docSections()
+	// sects := p.docSections()
```
Comment (emidoots) [inline contentprovider.go]: Do we want this left here commented out? Why?

File: contentprovider.go [hunk 5]
```diff
@@ -270,18 +270,22 @@ func matchScore(secs []DocumentSection, m *LineMatch) float64 {
...
-		if sec != nil {
-			startMatch := sec.Start == f.Offset
-			endMatch := sec.End == f.Offset+uint32(f.MatchLength)
-			if startMatch && endMatch {
-				score += scoreSymbol
-			} else if startMatch || endMatch {
-				score += (scoreSymbol + scorePartialSymbol) / 2
-			} else {
-				score += scorePartialSymbol
+		// Disable scoring based on symbol boundaries.
```
Comment (emidoots) [inline contentprovider.go]: Is the intent that we would bring this back in the future at some point? Would removing it outright be better? (can always look back at git history)

File: index_test.go [hunk 6]
```diff
@@ -1487,7 +1495,7 @@ func TestIOStats(t *testing.T) {
 	res := searchForTest(t, b, q)
 
 	// 4096 (content) + 2 (overhead: newlines or doc sections)
```
Comment (emidoots) [inline index_test.go]: This comment is no longer accurate

File: indexdata.go [hunk 7]
```diff
@@ -94,14 +94,26 @@ type symbolData struct {
...
+	symIndex   []byte
 	// symKindContent is an enum of sym.Kind and sym.ParentKind
 	symKindContent []byte
 	symKindIndex   []uint32
-	//symMetadata is [4]uint32 Sym Kind Parent ParentKind
+	// symMetadata is [4]uint32 Sym Kind Parent ParentKind
 	symMetaData []byte
 }
 
+func (d *symbolData) parent(i uint32) []byte {
```
Comment (emidoots) [inline indexdata.go:114]: Can you add a docstring to this function describing what it means?

- What is a parent in this context?
- What exactly is `i` referencing?
- What is the returned byte array?

File: read.go [hunk 8]
```diff
@@ -178,9 +177,23 @@ func (r *reader) readIndexData(toc *indexTOC) (*indexData, error) {
...
-		d.symIndex = toc.symbolMap.relativeIndex()
-		d.symKindIndex = toc.symbolKindMap.relativeIndex()
+	if d.metaData.IndexFormatVersion == 16 {
+		d.symbols.symKindIndex = toc.symbolKindMap.relativeIndex()
+		d.fileEndSymbol, err = readSectionU32(d.file, toc.fileEndSymbol)
+		if err != nil {
+			return nil, err
+		}
+
+		for sect, blob := range map[simpleSection]*[]byte{
```
Comment (emidoots) [inline read.go:189]: Why are these `*[]byte`? A slice is already a pointer to the underlying array value, so you can just use `[]byte` here and avoid the extra dereferencing this incurs.

Comment (emidoots) [reply read.go:189]: Ah, I misread and see now the code here is actually altering the value.

Comment (emidoots) [inline read.go:189]:
```suggestion
		// Call readSectionBlob on each section key, and store the result in
		// the blob value.
		for sect, blob := range map[simpleSection]*[]byte{
```

Comment (emidoots) [reply read.go:189]: This would've resolved my confusion when reading the code initially, I think.

Comment (kzh) [reply read.go:189]: This is because we're assigning to fields of the struct `d.symbols`. If it was a `[]byte`, we would only be able to assign to the local variable `blob`.

File: section.go [hunk 9]
```diff
@@ -159,3 +159,14 @@ func (s *compoundSection) relativeIndex() []uint32 {
 	}
 	return ri
 }
+
+type lazyCompoundSection struct {
+	compoundSection
+}
+
+func (s *lazyCompoundSection) read(r *reader) error {
+	if err := s.data.read(r); err != nil {
```
Comment (emidoots) [inline section.go:170]:
```suggestion
	// We do the same thing compoundSection.read does, except we don't read the
	// offsets.
	if err := s.data.read(r); err != nil {
```

File: section.go [hunk 10]
```diff
@@ -159,3 +159,14 @@ func (s *compoundSection) relativeIndex() []uint32 {
 	}
 	return ri
 }
+
+type lazyCompoundSection struct {
```
Comment (emidoots) [inline section.go:163]: I see that this type is no longer responsible for parsing the `offsets` out of the underlying data and instead that logic is now happening iteratively in `(*symbolData).parent`.

I think a more clear way to make this happen would be via methods on this type to iteratively consume the offsets that `(*symbolData).parent` would call into. What do you think?

File: contentprovider.go [hunk 11]
```diff
@@ -154,9 +154,8 @@ func (p *contentProvider) fillMatches(ms []*candidateMatch) []LineMatch {
 		result = p.fillContentMatches(ms)
 	}
 
-	sects := p.docSections()
```
Comment (keegancsmith) [inline contentprovider.go:157]: can you add a comment here explaining the divergence from upstream / why we pass in nil. This will help give context if we encounter a merge conflict here.

File: contentprovider.go [hunk 12]
```diff
@@ -218,8 +217,16 @@ func (p *contentProvider) fillContentMatches(ms []*candidateMatch) []LineMatch {
...
 				LineOffset:  int(m.byteOffset) - lineStart,
 				MatchLength: int(m.byteMatchSz),
-				SymbolInfo:  m.symbolInfo,
 			}
+			if m.symbol {
+				start := p.id.fileEndSymbol[p.idx]
+				fragment.SymbolInfo = p.id.symbols.data(start + m.symbolIdx)
+				if fragment.SymbolInfo != nil {
+					sec := p.docSections()[m.symbolIdx]
+					fragment.SymbolInfo.Sym = string(data[sec.Start:sec.End])
```
Comment (keegancsmith) [inline contentprovider.go:226]: maybe for follow-up PR: Should we make SymbolInfo fields all `[]byte`? I see above for `.Line` we use a []byte which avoids the need to copy the data into a string.

File: contentprovider.go [hunk 13]
```diff
@@ -270,18 +277,6 @@ func matchScore(secs []DocumentSection, m *LineMatch) float64 {
 			score = scorePartialWordMatch
 		}
 
-		sec := findSection(secs, f.Offset, uint32(f.MatchLength))
```
Comment (keegancsmith) [inline contentprovider.go:273]: comment for why it is deleted to help merging.

File: indexbuilder.go [hunk 14]
```diff
@@ -368,12 +368,11 @@ func (b *IndexBuilder) symbolKindID(t string) uint32 {
 func (b *IndexBuilder) addSymbols(symbols []*Symbol) {
 	for _, sym := range symbols {
 		b.symMetaData = append(b.symMetaData,
-			b.symbolID(sym.Sym),
+			0,
```
Comment (keegancsmith) [inline indexbuilder.go:375]: why are we still storing this? For backwards compatibility on v16?

File: indexdata.go [hunk 15]
```diff
@@ -167,6 +188,7 @@ func (d *indexData) memoryUse() int {
 		d.boundaries, d.fileNameIndex,
 		d.runeOffsets, d.fileNameRuneOffsets,
 		d.fileEndRunes, d.fileNameEndRunes,
+		d.fileEndSymbol, d.symbols.symKindIndex,
```
Comment (keegancsmith) [inline indexdata.go:191]: nice catch

File: matchiter.go [hunk 16]
```diff
@@ -35,7 +35,8 @@ type candidateMatch struct {
 	byteOffset  uint32
 	byteMatchSz uint32
 
-	symbolInfo *Symbol
+	symbol    bool
+	symbolIdx uint32
```
Comment (keegancsmith) [inline matchiter.go]: we can do better with alignment and save space. Check the proposed place to put these fields, just after the bools at the top of the strucT: http://golang-sizeof.tips/?t=c3RydWN0IHsKCWNhc2VTZW5zaXRpdmUgYm9vbAoJZmlsZU5hbWUgICAgICBib29sCgogICAgc3ltYm9sICAgIGJvb2wKICAgIHN5bWJvbElkeCB1aW50MzIKCglzdWJzdHJCeXRlcyAgIFtdYnl0ZQoJc3Vic3RyTG93ZXJlZCBbXWJ5dGUKCglmaWxlIHVpbnQzMgoKCS8vIE9mZnNldHMgYXJlIHJlbGF0aXZlIHRvIHRoZSBzdGFydCBvZiB0aGUgZmlsZW5hbWUgb3IgZmlsZSBjb250ZW50cy4KCXJ1bmVPZmZzZXQgIHVpbnQzMgoJYnl0ZU9mZnNldCAgdWludDMyCglieXRlTWF0Y2hTeiB1aW50MzIKCn0K

File: matchtree.go [hunk 17]
```diff
@@ -213,29 +210,29 @@ func (t *symbolSubstrMatchTree) prepare(doc uint32) {
 		fileStart = t.fileEndRunes[doc-1]
 	}
 
-	for len(t.sections) > 0 && t.sections[0].Start < fileStart {
-		t.sections = t.sections[1:]
-		t.secID++
+	var sections []DocumentSection
+	if len(t.sections) > 0 {
+		sections = t.sections[t.fileEndSymbol[doc]:t.fileEndSymbol[doc+1]]
```
Comment (keegancsmith) [inline matchtree.go:215]: does doc+1 always exist?

File: matchtree.go [hunk 18]
```diff
@@ -213,29 +210,29 @@ func (t *symbolSubstrMatchTree) prepare(doc uint32) {
...
 	}
+	secIdx := 0
 
-	// TODO we can use fileEndSymbol to skip
 	trimmed := t.current[:0]
-	for len(t.sections) > 0 && len(t.current) > 0 {
+	for len(sections) > secIdx && len(t.current) > 0 {
```
Comment (keegancsmith) [inline matchtree.go:220]: IMPORTANT: I didn't realise this when I first reviewed, but you are doing far too much work in prepare. Prepare should pretty much be O(1). You should only do this work in matches.

IE if you do a search like `f:foo sym:bar` Prepare will run for every document. However, matches won't run with a high enough cost for symbol since f:foo matchtree will rule out most documents.

So essentially this whole function should be moved into matches, and prepare should only reset fields in the struct.

File: read.go [hunk 19]
```diff
@@ -178,9 +177,25 @@ func (r *reader) readIndexData(toc *indexTOC) (*indexData, error) {
...
+	if d.metaData.IndexFormatVersion == 16 {
+		d.symbols.symKindIndex = toc.symbolKindMap.relativeIndex()
+		d.fileEndSymbol, err = readSectionU32(d.file, toc.fileEndSymbol)
+		if err != nil {
+			return nil, err
+		}
+
+		// Call readSectionBlob on each section key, and store the result in
+		// the blob value.
+		for sect, blob := range map[simpleSection]*[]byte{
```
Comment (keegancsmith) [inline read.go:189]: I'm guessing this won't be the case, but is there any possibility of two simpleSections have the same value? The downside of using a map is keys are unique.

**Review by keegancsmith — APPROVED** [review]
some important inline feedback, but I trust you understand it so I don't need to review the update. So accepting to unblock landing.

File: matchiter.go (hunk 16 above)
Comment (kzh) [reply matchiter.go]: Interesting, I didn't know this was the case. I figure this would be an optimization that the Go compiler handles.

File: indexbuilder.go (hunk 14 above)
Comment (kzh) [reply indexbuilder.go:375]: I hope to remove the first uint32 entirely in the future. This would most likely require a change in the feature version, thus reindexing. For now, setting it to `0` will remove the redundant storage of the symbol names as well as avoid reindexing.

File: read.go (hunk 19 above)
Comment (kzh) [reply read.go:189]: I don't think so. Simple sections should refer to unique locations (byte offset + sz) in the index format.

File: matchtree.go (hunk 17 above)
Comment (kzh) [reply matchtree.go:215]: Yes, `len(t.fileEndSymbol)` will always be one greater than the number of documents in the case that there are ctags information stored (`len(t.sections) > 0`).

File: read.go (hunk 19 above)
Comment (emidoots) [reply read.go:189]: A map is probably slower than a slice here BTW. I am speculating but maps usually require many more allocations and computations and I don't imagine the Go compiler can optimize this down to something simpler.

That said, if this isn't a critical section of code it doesn't matter.

File: matchtree.go (hunk 18 above)
Comment (kzh) [reply matchtree.go:220]: Gotcha, will change. I initially placed it here to resemble upstream zoekt (which does this work in prepare).

~The prepare of `symbolSubstrMatchTree` has the same time complexity and reponsibility as upstream `substrMatchTree`'s prepare. They both pre-populate potential candidates. The difference is that symbolSubstrMatchTree trims that list as well, but this operation is also O(n).~

File: read.go (hunk 19 above)
Comment (kzh) [reply read.go:189]: I can switch to a slice. Upstream Zoekt has a similar piece of code further down this function.
```go
for sect, dest := range map[simpleSection]*[]uint32{
	toc.subRepos:        &d.subRepos,
	toc.runeOffsets:     &d.runeOffsets,
	toc.nameRuneOffsets: &d.fileNameRuneOffsets,
	toc.nameEndRunes:    &d.fileNameEndRunes,
	toc.fileEndRunes:    &d.fileEndRunes,
} {
	if blob, err := d.readSectionBlob(sect); err != nil {
		return nil, err
	} else {
		*dest = fromSizedDeltas(blob, nil)
	}
}
```
This reading logic happens once when a repository is loaded in (and reindexed). This would affect the speed at which the indexed-search pod's `zoekt-webserver` container becomes ready I think.

File: matchtree.go (hunk 18 above)
Comment (kzh) [reply matchtree.go:220]: ~I might be misunderstanding Zoekt, but I don't think prepare runs on every document. It runs only on documents returned by nextDoc (so only documents that have foo in its name and bar in its content) so there shouldn't be any extraneous work done here.~

https://sourcegraph.com/github.com/sourcegraph/zoekt/-/blob/eval.go#L165

Screen Shot 2019-09-13 at 4 25 51 PM
Screen Shot 2019-09-13 at 4 26 21 PM

```go
// An expression tree coupled with matches. The matchtree has two
// functions:
//
// * it implements boolean combinations (and, or, not)
//
// * it implements shortcuts, where we skip documents (for example: if
// there are no trigram matches, we can be sure there are no substring
// matches). The matchtree iterates over the documents as they are
// ordered in the shard.
//
// The general process for a given (shard, query) is
//
// - construct matchTree for the query
//
// - find all different leaf matchTrees (substring, regexp, etc.)
//
// in a loop:
//
//   - find next doc to process using nextDoc
//
//   - evaluate atoms (leaf expressions that match text)
//
//   - evaluate the tree using matches(), storing the result in map.
//
//   - if the complete tree returns (matches() == true) for the document,
//     collect all text matches by looking at leaf matchTrees
//
type matchTree interface {
	docIterator

	// returns whether this matches, and if we are sure.
	matches(cp *contentProvider, cost int, known map[matchTree]bool) (match bool, sure bool)
}

// A docIterator iterates over documents in order.
type docIterator interface {
	// provide the next document where we can may find something
	// interesting.
	nextDoc() uint32

	// clears any per-document state of the docIterator, and
	// prepares for evaluating the given doc. The argument is
	// strictly increasing over time.
	prepare(nextDoc uint32)
}
```

File: read.go (hunk 19 above)
Comment (emidoots) [reply read.go:189]: Ahh, got it, probably not a huge perf benefit then.

File: matchtree.go (hunk 18 above)
Comment (keegancsmith) [reply matchtree.go:220]: Yes, `prepare` is only called on `nextDoc` which means it should generally only be prepared on documents which actually have matches. But just because a matchtree returns a doc from nextDoc doesn't imply it actually matches. As a stark example try `f:go$` then nextDoc can't rely on trigram indexes.

But yeah given our trigram indexes work so well this isn't as important as I thought! Sorry for delaying this PR. Rather ship this idea in a follow up (and possibly upstream?)

File: indexbuilder.go (hunk 14 above)
Comment (keegancsmith) [reply indexbuilder.go:375]: Cool, maybe add a TODO explaining that. IE we can remove it, it will just require bumping feature version.

**Review by keegancsmith — APPROVED** [review]

File: matchtree.go (hunk 18 above)
Comment (kzh) [reply matchtree.go:220]: Thanks for explaining! I'll follow up on this in a separate PR.
